src/qt3utils/applications/controllers/princeton_spectrometer.py

```diff
@@ -0,0 +1,142 @@
+import numpy as np
+import tkinter as tk
+import logging
+from typing import Tuple
+
+import qt3utils.datagenerators.princeton as princeton
+
+
+class QT3ScanPrincetonSpectrometerController:
+    """
+    Implements qt3utils.applications.qt3scan.interface.QT3ScanSpectrometerDAQControllerInterface
+    """
+    def __init__(self, logger_level: int):
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(logger_level)
+
+        self.spectrometer = princeton.SpectrometerDataAcquisition()
+        self.last_config_dict = {}
+
+        self.wave_start = None
+        self.wave_end = None
+        self.last_measured_spectrum = None
+        self.last_wavelength_array = None
+
+    @property
+    def clock_rate(self) -> float:
+        try:
+            _t = self.spectrometer.exposure_time / 1000.0  # Converting from milliseconds to seconds.
+        except Exception as e:
+            self.logger.error(e)
+            _t = 2  # TODO: better default behavior. Should this be -1? 1? or should Spectrometer be changed.
+        return 1.0 / _t
+
+    def start(self) -> None:
+        """
+        Nothing to be done in this method. All acquisition is happening in the "sample_spectrum" method.
+        """
+        self.logger.debug('calling QT3ScanPrincetonSpectrometerController start')
+
+    def stop(self) -> None:
+        """
+        Implementations should do necessary steps to stop acquiring data.
+        """
+        self.spectrometer.stop_scan()
+        self.logger.debug('calling QT3ScanPrincetonSpectrometerController stop')
+
+    def close(self) -> None:
+        self.spectrometer.finalize()
+
+    def sample_spectrum(self) -> Tuple[np.ndarray, np.ndarray]:
+        self.last_measured_spectrum, self.last_wavelength_array = (self.spectrometer.acquire_step_and_glue([self.wave_start, self.wave_end]))
```

Comment:
why acquire_step_and_glue?

Reply:
Just a note here that evidently we acquire data differently in the lab (something that I was not aware of). But we have spoken about this